Sprint 10

bakgeldia opened the pull request:
Задача спринта #10:
1. Приложение зарегистрировано на сайте Unsplash.
2. Экран авторизации содержит логотип приложения и кнопку «Войти».
3. Для входа в приложение пользователь авторизуется через OAuth2.0
4. В проекте создан файл Constants.swift, в нём есть константы accessKey, secretKey, redirectURI, accessScope и defaultBaseURL
5. Сверстан контроллер WebView по ТЗ.
6. Реализован KVO
7. Проверка необходимости авторизации выделена в отдельный контроллер SplashViewController
8. Добавлена структура для декодинга JSON-ответа от POST-запроса авторизации Unsplash — OAuthTokenResponseBody
9. Создан новый класс OAuth2Service с нужными функциями
10. Добавлено логирование ошибок в консоль 
11. Использовано безопасное извлечение значений опционалов и применен do-catch для throwable-методов.

bakgeldia committed:
- Реализованы контроллеры AuthViewController и WebViewViewController.
- Написан запрос для получения кода авторизации.
- С помощью навигационного делегата получение уведомления о действиях пользователя и обрабатывание их.

bakgeldia committed:
Используя KVO, реализовано обновление в прогрессе загрузки веб-станицы авторизации

bakgeldia committed:
Получение access_token'а через POST-запрос

bakgeldia committed:
Реализован SplashViewController, который является стартовой точкой приложения. Данный контроллер проверяет, сохранен ли Bearer Token в UserDefaults. Если так, то пользователя кидает сразу в ленту, иначе показывается AuthViewController для дальнейшей авторизации.

andrew-kuzn-yndx commented on ImageFeed/Auth/AuthViewController.swift:
```diff
@@ -0,0 +1,80 @@
+//    self.dismiss(animated: true)
+//case .failure(let error):
+//    print("Error fetching token: \(error)")
+//}
```
Старайся не оставлять такое, правилом хорошего тона является чистить код от лишнего перед сдачей работы. В основную ветку проекта должен попадать только чистый код без лишних принтов и закомментированных блоков)

andrew-kuzn-yndx commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,120 @@
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        webView.removeObserver(self, forKeyPath: #keyPath(WKWebView.estimatedProgress), context: nil)
```
Молодец, что не забыл удалить Observer!)

andrew-kuzn-yndx commented on ImageFeed/Auth/WebViewViewController.swift:
```diff
@@ -0,0 +1,120 @@
+        if
+            let url = navigationAction.request.url,
+            let urlComponents = URLComponents(string: url.absoluteString),
+            urlComponents.path == "/oauth/authorize/native",
```
Это тоже можно поместить в WebViewConstants

bakgeldia committed:
Мини-исправления после code-review:
- Удалены лишние комментарий
- Путь "/oauth/authorize/native" помещен в константу
